File: app/src/main/java/com/soobinpark/appcraft/readingbible/data/parser/LfaBibleFileParser.kt
package com.soobinpark.appcraft.readingbible.data.parser

import com.soobinpark.appcraft.readingbible.domain.model.BibleBook
import com.soobinpark.appcraft.readingbible.domain.model.BibleCatalog
import com.soobinpark.appcraft.readingbible.domain.model.BibleVerse
import com.soobinpark.appcraft.readingbible.domain.model.BibleVersion
import java.io.File
import java.util.zip.ZipFile

class LfaBibleFileParser : BibleFileParser {
    override suspend fun readChapter(
        version: BibleVersion,
        book: BibleBook,
        chapter: Int,
    ): List<BibleVerse> {
        val root = version.fileRoot ?: return emptyList()
        val archive = File(root, "${version.code}.lfa")
        if (!archive.exists()) return emptyList()

        val bookNumber = BibleCatalog.bookNumber(book.index)
        val chapterNeedle = " $chapter:"
        val entryName = "${version.code}${bookNumber}_$chapter.lfb"

        return ZipFile(archive).use { zip ->
            val entry = zip.getEntry(entryName) ?: return emptyList()
            zip.getInputStream(entry).bufferedReader().useLines { lines ->
                lines.mapNotNull { line ->
                    parseVerseLine(version.code, book.index, chapter, bookNumber, chapterNeedle, line)
                }.toList()
            }
        }
    }
}

File: ios/Sources/Data/Parsers/LfaBibleFileParser.swift
import Foundation
import ZIPFoundation

struct LfaBibleFileParser: Sendable {
    func readChapter(version: BibleVersion, book: BibleBook, chapter: Int) throws -> [BibleVerse] {
        let archiveURL = version.fileRoot.appendingPathComponent("\(version.code).lfa")
        guard FileManager.default.fileExists(atPath: archiveURL.path) else { return [] }

        let bookNumber = BibleCatalog.bookNumber(forBookIndex: book.index)
        let chapterNeedle = " \(chapter):"
        let entryName = "\(version.code)\(bookNumber)_\(chapter).lfb"

        guard let archive = Archive(url: archiveURL, accessMode: .read),
              let entry = archive.first(where: {
                  $0.path.caseInsensitiveCompare(entryName) == .orderedSame
              })
        else { return [] }

        var data = Data()
        _ = try archive.extract(entry) { chunk in data.append(chunk) }
        let text = BibleTextEncoding.decodeLfaText(data)

        return text.components(separatedBy: .newlines).compactMap { line in
            VerseLineParser.parse(
                versionCode: version.code,
                bookIndex: book.index,
                chapter: chapter,
                bookNumber: bookNumber,
                chapterNeedle: chapterNeedle,
                line: line
            )
        }
    }
}
